### PR TITLE
chore: refactor agent stats streaming

### PR DESCRIPTION
**Pull request by coadler**
Agents no longer use a websocket for streaming stats, and instead use the returned interval to determine when to send the next stat.

Network stats are retrieved directly from Tailscale, getting rid of the need to wrap network connections.

### CODE REVIEW

File: coderd/coderd.go [hunk 1]
```diff
@@ -493,7 +493,9 @@ func New(options *Options) *API {
 				r.Get("/gitauth", api.workspaceAgentsGitAuth)
 				r.Get("/gitsshkey", api.agentGitSSHKey)
 				r.Get("/coordinate", api.workspaceAgentCoordinate)
-				r.Get("/report-stats", api.workspaceAgentReportStats)
+				r.Post("/report-stats", api.workspaceAgentReportStats)
+				// DEPRECATED
+				r.Get("/report-stats", api.workspaceAgentReportStatsWebsocket)
```
Comment (deansheather) [inline coderd/coderd.go:498]:
```suggestion
				// DEPRECATED in favor of the POST endpoint above.
				// TODO: remove in January 2023
				r.Get("/report-stats", api.workspaceAgentReportStatsWebsocket)
```

File: coderd/database/queries/workspaces.sql [hunk 2]
```diff
@@ -8,6 +8,35 @@ WHERE
...
+					workspace_resources.id = (
+						SELECT
+							resource_id
+						FROM
+							workspace_agents
+						WHERE
+							workspace_agents.id = @agent_id
+					)
+			)
+	);
```
Comment (deansheather) [inline coderd/database/queries/workspaces.sql:38]: ouch...

File: coderd/workspaceagents.go [hunk 3]
```diff
@@ -755,31 +755,69 @@ func convertWorkspaceAgent(derpMap *tailcfg.DERPMap, coordinator tailnet.Coordin
...
+		Payload:     json.RawMessage("{}"),
+	})
+	if err != nil {
+		httpapi.InternalServerError(rw, err)
+		return
+	}
+
+	err = api.Database.UpdateWorkspaceLastUsedAt(ctx, database.UpdateWorkspaceLastUsedAtParams{
+		ID:         workspace.ID,
+		LastUsedAt: now,
```
Comment (deansheather) [inline coderd/workspaceagents.go:799]: would this not cause every workspace to get bumped every 5 minutes regardless of activity as long as it had ever received and transmitted at least 1 byte?

Comment (deansheather) [reply coderd/workspaceagents.go:799]: Oh, do the counters get reset each push?

File: coderd/workspaceagents.go [hunk 4]
```diff
@@ -755,31 +755,69 @@ func convertWorkspaceAgent(derpMap *tailcfg.DERPMap, coordinator tailnet.Coordin
 func (api *API) workspaceAgentReportStats(rw http.ResponseWriter, r *http.Request) {
```
Comment (deansheather) [inline coderd/workspaceagents.go:755]: This should have tests that check the stat gets updated in the db and that the last used at gets updated properly

**Review by deansheather — APPROVED** [review]

File: coderd/workspaceagents.go (hunk 3 above)
Comment (coadler) [reply coderd/workspaceagents.go:799]: Yeah the counters get reset on each push, so you're only sending the activity from that period.

**Pull request merged**